### PR TITLE
Refactoring of `addon-info`

### PR DESCRIPTION
**Pull request by usulpro**
Issue: https://github.com/storybooks/storybook/issues/1147#issuecomment-306409320 and a discussion below

since we're slowly moving from .addWithX addons to more wise solutions, we need an appropriate version of the addon

## What I did

1. Added `withInfo()` function.

accessible via:
```js
import { withInfo } from '@storybook/addon-info';
```

API:

```js
infoFn(context => { /* return <Story /> */ }) = withInfo(storyInfo, options)
```

Example:

```js
.add('with new info', withInfo('info addon')(context =>
      <div>
        click the <InfoButton /> label in top right for info about "{context.story}"
      </div>
))
```

2 Added deprecation warning to `.addWithInfo`

3 Added examples to cra-kitchen-sink:

- new `withInfo()` API

- `withInfo()` and `addonNotes()` composition

## How to test

- run cra-kitchen-sink 

- run tests

### CODE REVIEW

File: addons/info/src/index.js [hunk 1]
```diff
@@ -29,58 +54,61 @@ const defaultMarksyConf = {
...
 
-      return (
-        <Story {...props}>
-          {storyFn(context)}
-        </Story>
-      );
-    });
+export default {
+  addWithInfo(storyName, info, storyFn, _options) {
+    return this.add(storyName, warning(withInfo(info, _options)(storyFn)));
```
Comment (shilman) [inline addons/info/src/index.js]: Let's just deprecate `addWithInfo`?

**Review by shilman — CHANGES_REQUESTED** [review]
Thanks so much for doing this. I think it will help Storybook's addon ecosystem a lot!

I requested some changes around deprecation.

I also have a question. Is the `context` argument to stories a convention in Storybook that `addon-info` was stripping out? Or is it something that we are adding recently?

**Comment by shilman** [comment]
@alexandrebodin @ndelangen @tmeasday mind reviewing this? it's a huge step towards cleaning up the addon ecosystem, but let's make sure we all agree on the specifics before we charge ahead!

**Comment by UsulPro** [comment]
@shilman Yes, passing `context` to `storyFn()` (and therefore to `decoratorFn()`) is existing API, it was introduced already in v1.31.0. So I made a `withInfo()` with a support of this API as well.

I tried to use `util-deprecate` it's convenient but I could not get it to behave as I need for several reasons (not critical in general):

- It does not support passing arguments into warning messages while I wanted to show the specified story and storyKind in the notification

- If I wrap `addWithInfo` with `util-deprecate` it'll change the name of the function so it will not be available as `storiesOf().addWithInfo` (possible could be a workaround for that though). And I wanted the warning to be shown only when you select the story, not in the beginning.

Do you think that we should intend to use `util-deprecate` in all cases?

File: addons/info/src/index.js [hunk 2]
```diff
@@ -2,6 +2,31 @@ import React from 'react';
...
+  const deprecated = msg => {
+    if (!warned) {
+      logger.warn(msg);
+      warned = true;
+    }
+  };
+  return deprecated;
+}
+
+const showWaring = deprecate();
```
Comment (UsulPro) [inline addons/info/src/index.js]: OH! Sure!!!

File: addons/info/src/index.js (hunk 1 above)
Comment (UsulPro) [inline addons/info/src/index.js]: see my answer in a comment below, please!

**Comment by shilman** [comment]
@UsulPro personal preference would be to just use `util-deprecate` across the board and keep the code size down, although you won't be able to access the arguments in the message.

As for changing the name of the function, it shouldn't be a problem:
```
export default {
  addWithInfo: deprecate(deprecatedAddWtihInfo, message),
  ...
}
```

**Comment by tmeasday** [comment]
Looks great @UsulPro, great work!

@shilman what specifics about the addon ecosystem were you concerned about?

I think exporting a "decorator" like this `withInfo` is the best way forward that fits with the current design. One minor thing: should we use [recompose's](https://github.com/acdlite/recompose) `compose` for combining decorators in our examples? It might be slightly simpler and easier to follow?

Also I noticed that the knobs example isn't working on this branch. Is this a known problem?

**Review by tmeasday — APPROVED** [review]

**Comment by shilman** [comment]
@tmeasday I'm only concerned insofar as changes to the API are expensive and I want to make sure we're all bought in. I think we're in total agreement that `withX` decorators is the way to go, but I just want to make sure.

**Review by shilman — APPROVED** [review]

**Comment by shilman** [comment]
@tmeasday Good catch on the knobs. Tracking here: https://github.com/storybooks/storybook/issues/1464

**Comment by shilman** [comment]
@UsulPro does the change to the default props change behavior in meaningful ways that could break things for existing users?

**Comment by shilman** [comment]
@UsulPro CI is breaking. do you mind fixing this and then I'll merge:
```
Summary of all failing tests
 FAIL  addons/info/src/index.test.js
  ● addon Info › should render with missed info
    No story defining function has been specified
      
      at addInfo (addons/info/src/index.js:98:91)
      at Object.<anonymous> (addons/info/src/index.test.js:53:19)
          at Promise (<anonymous>)
      at Promise.resolve.then.el (node_modules/p-map/index.js:42:16)
          at <anonymous>
      at process._tickCallback (internal/process/next_tick.js:169:7)
```

File: addons/info/src/components/PropVal.js [hunk 3]
```diff
@@ -116,11 +116,11 @@ export default function PropVal(props) {
 }
 
 PropVal.defaultProps = {
-  val: null, // eslint-disable-line
-  maxPropObjectKeys: 0,
-  maxPropArrayLength: 0,
-  maxPropStringLength: 0,
-}
+  val: null,
+  maxPropObjectKeys: 3,
```
Comment (shilman) [inline addons/info/src/components/PropVal.js]: Default values should be "show everything"

**Review by shilman — CHANGES_REQUESTED** [review]

**Review by shilman — APPROVED** [review]

**Pull request merged**